qa/task-cancellation/src/main/java/org/elasticsearch/plugin/tasks/TransportTestAction.java

```diff
@@ -0,0 +1,175 @@
+import org.elasticsearch.action.ActionListenerResponseHandler;
+import org.elasticsearch.action.ActionType;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.GroupedActionListener;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.util.concurrent.AbstractRunnable;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.tasks.TaskCancelledException;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.Transport;
+import org.elasticsearch.transport.TransportActionProxy;
+import org.elasticsearch.transport.TransportRequestOptions;
+import org.elasticsearch.transport.TransportService;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+
+public class TransportTestAction extends HandledTransportAction<TestRequest, TestResponse> {
+    static final Logger logger = LogManager.getLogger(TransportTestAction.class);
+    static final ActionType<TestResponse> ACTION = new ActionType<>("internal::test_action", TestResponse::new);
+    final TransportService transportService;
+    final Map<String, CountDownLatch> latches = ConcurrentCollections.newConcurrentMap();
+    final ClusterService clusterService;
+
+    @Inject
+    public TransportTestAction(TransportService transportService, ActionFilters actionFilters, ClusterService clusterService) {
+        super(ACTION.name(), transportService, actionFilters, TestRequest::new, ThreadPool.Names.GENERIC);
+        this.transportService = transportService;
+        this.clusterService = clusterService;
+        TransportActionProxy.registerProxyAction(transportService, ACTION.name(), TestResponse::new);
+    }
+
+    @Override
+    protected void doExecute(Task task, TestRequest request, ActionListener<TestResponse> origListener) {
+        ActionListener<TestResponse> groupedListener = new GroupedActionListener<>(
+            ActionListener.map(origListener, resp -> new TestResponse(resp.stream().allMatch(AcknowledgedResponse::isAcknowledged))),
+            request.targets.size()
+        );
+        boolean executeLocally = false;
+        for (TestRequest.Target target : request.targets) {
+            if (target.nodeId.equals(transportService.getLocalNode().getId())) {
+                executeLocally = true;
+            } else {
+                dispatchSubRequest(task, request, target, groupedListener);
+            }
+        }
+        if (executeLocally) {
+            executeRequest((BlockingCancellableTask) task, request.id, groupedListener);
```

Comment:
why execute these last and not at the place where you do `executeLocally = true`?

Reply:
`executeRequest` was executed blockingly before. I pushed https://github.com/elastic/elasticsearch/pull/59828/commits/5c240346c6e7ffde794f7a52d0d4757d3df57e62